### PR TITLE
Jhbate/datamodel migration process

### PR DESCRIPTION
**Pull request by jh-bate**
@darinkrauss @jebeck 

Not right yet but a start and hopefully you get the idea of what I am trying to achieve. Feel free to suggest alternatives of better places to keep these notes

### CODE REVIEW

File: _docs/data-model/migrations/migrations.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+layout: defaults
+title: Data Model Migrations
+published: true
+---
+
+### _schemaVersion=0
+
+This migration applies to change introducted as part of the "UTC Bootstrap" ADD_REFERENCE_HERE changes and introduces a generic [`_schemaVersion`](v1#versioning-and-updates) to our data model.
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: "change introducted" to "changes introduced"

File: _docs/data-model/migrations/migrations.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+**Check**
+
+In the mongo console check to see if you need to apply the migration by finding how many records don't have the `_schemaVersion`
+
+```
+> db.deviceData.find({"_schemaVersion": {"$exists”:false}}).count()
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Need "> use streams" first.

Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Also, the quote after $exists is a smart quote.  Copy and paste fail to work. Needs to be a plain double quote.

File: _docs/data-model/migrations/migrations.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+```
+49903
+```
+
+**Backup**
+
+In mongo backup your existing data collection by performing a copyTo
+
+```
+$ mongoexport --db data --collection deviceData --out deviceData_PreSchemaVersion.json
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Is it the "streams" database?  Or do I have something misconfigured on my machine.  I have the following mongo databases: 

```
> show dbs
data	(empty)
gate	0.203125GB
hydrophone	0.203125GB
local	0.078125GB
messages	0.203125GB
streams	3.9521484375GB
test	(empty)
test_ingestion	0.203125GB
test_streams	0.203125GB
user	0.203125GB
```

File: _docs/data-model/migrations/migrations.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+You should revieve feedback that the update has succeeded similar to the following
+
+```
+WriteResult({ "nMatched" : 49903, "nUpserted" : 0, "nModified" : 49903 })
+```
+
+```
+> db.deviceData.find({"_schemaVersion": {"$exists”:false}}).count()
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Might want a comment about this should return 0.

File: _docs/data-model/migrations/migrations.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+**Rollback** (if required)
+
+If you want run the process again or rollback for any reason then do the following
+
+```
+$ mongoimport --db data --collection deviceData --type json --file ./deviceData_PreSchemaVersion.json --jsonArray
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: As above: data db or streams db?

File: _docs/data-model/migrations/migrations.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+Once successful you will see a count of the records that have been copied back to your original collection.
+
+```
+$ imported 49903 records
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Does this only import and replace existing records at the time of the export?  If so, what about if there are any records created after the export?  Do those get dropped?

**Comment by darinkrauss** [comment]
Just some comments about the details, but overall looks good.

File: _docs/data-model/migrations/migrations.md (hunk 5 above)
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Actually, when I tried this, I got:

```
Wed Jul  8 13:31:01.754 exception:JSONArray file too large
Wed Jul  8 13:31:01.786 warning: log line attempted (16384k) over max size(10k), printing beginning and end ... { "dev
```

**Comment by darinkrauss** [comment]
Actually, I just finished testing the mongoimport and ran into an error.  See the comment inline.

**Comment by jh-bate** [comment]
@darinkrauss I think I have responded to all the items you mention above in the review

File: _docs/data-model/migrations/migrations.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+**Apply**
+
+On the collection you want to update run the following
+
+```
+> use data
+> db.deviceData.update({_schemaVersion: {"$exists":false}},{$set:{_schemaVersion:0}},{multi:true})
+```
+
+You should revieve feedback that the update has succeeded similar to the following.
```
Comment (darinkrauss) [inline _docs/data-model/migrations/migrations.md]: Receive

**Comment by darinkrauss** [comment]
One silly typo, but :shipit:

**Pull request merged**